relayer/chains/cosmos/cosmos_chain_processor.go

```diff
@@ -40,9 +60,147 @@ func (ccp *CosmosChainProcessor) SetPathProcessors(pathProcessors processor.Path
 	ccp.pathProcessors = pathProcessors
 }
 
+// latestHeightWithRetry will query for the latest height, retrying in case of failure.
+// It will delay by latestHeightQueryRetryDelay between attempts, up to latestHeightQueryRetries.
+func (ccp *CosmosChainProcessor) latestHeightWithRetry(ctx context.Context) (latestHeight int64, err error) {
+	return latestHeight, retry.Do(func() error {
+		latestHeightQueryCtx, cancelLatestHeightQueryCtx := context.WithTimeout(ctx, latestHeightQueryTimeout)
+		defer cancelLatestHeightQueryCtx()
+		var err error
+		latestHeight, err = ccp.chainProvider.QueryLatestHeight(latestHeightQueryCtx)
+		return err
+	}, retry.Context(ctx), retry.Attempts(latestHeightQueryRetries), retry.Delay(latestHeightQueryRetryDelay), retry.LastErrorOnly(true), retry.OnRetry(func(n uint, err error) {
+		ccp.log.Info(
+			"Failed to query latest height",
+			zap.String("chain_id", ccp.chainProvider.ChainId()),
+			zap.Uint("attempt", n+1),
+			zap.Uint("max_attempts", latestHeightQueryRetries),
+			zap.Error(err),
+		)
+	}))
+}
+
+type queryCyclePersistence struct {
+	latestHeight         int64
+	latestQueriedBlock   int64
+	minQueryLoopDuration time.Duration
+}
+
 // Run starts the query loop for the chain which will gather applicable ibc messages and push events out to the relevant PathProcessors.
 // The initialBlockHistory parameter determines how many historical blocks should be fetched and processed before continuing with current blocks.
 // ChainProcessors should obey the context and return upon context cancellation.
 func (ccp *CosmosChainProcessor) Run(ctx context.Context, initialBlockHistory uint64) error {
-	panic("not yet implemented")
+	// this will be used for persistence across query cycle loop executions
+	persistence := queryCyclePersistence{
+		minQueryLoopDuration: defaultMinQueryLoopDuration,
+	}
+
+	// Infinite retry to get initial latest height
+	for {
+		latestHeight, err := ccp.latestHeightWithRetry(ctx)
+		if err != nil {
```

Comment:
I think this block needs to check `errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded)` -- without that check, if the context is terminated during initial height retrieval, this method will loop forever.

Reply:
good catch, done.